rflx/generator/allocator.py

```diff
@@ -218,27 +235,38 @@ def _needs_allocation(type_: rty.Type) -> bool:
 
     def _allocate_slots(self) -> None:
         """Create memory slots for each construct in the session that requires memory."""
+        # pylint: disable=too-many-locals
         count: int = 0
 
-        def insert(loc: Optional[Location]) -> None:
+        def create_slot_with_size(size: int) -> int:
             nonlocal count
             count += 1
-            assert loc is not None
-            self._allocation_map[loc] = count
             self._all_slots.add(count)
+            self._size_map[count] = size
+            return count
+
+        def map_sloc_to_slot(loc: Optional[Location], slot: int) -> None:
+            assert loc is not None
+            self._allocation_map[loc] = slot
 
+        # global variables
         for d in self._session.declarations.values():
             if isinstance(d, decl.VariableDeclaration) and self._needs_allocation(d.type_):
-                insert(d.location)
-                self._global_slots.add(count)
+                slot = create_slot_with_size(self._session.buffer_sizes.get_size(d.identifier))
+                map_sloc_to_slot(d.location, slot)
+                self._global_slots.add(slot)
 
-        global_count = count
+        # local variables
 
+        # get all allocation points and required sizes in a list (one element per state)
+        # of lists of tuples (loc, size)
+
+        state_allocation_list = []
         for s in self._session.states:
```

Comment:
The `too-many-locals` warning isn't needed anymore. I kind of like the structure of this function, so I wouldn't change unless you feel strongly about it.
Also, aren't comments a good thing? I would leave them in even if I were to pull out some code in a separate method.

Reply:
> Also, aren't comments a good thing?

It depends. Comments can be very helpful if they describe the *Why* (Why does the code exist?). But comments should not describe the *What*. The *What* should be apparent from the code itself.

Reply:
I'm aware of the Why/What distinction, but I don't fully agree. When the "What" is complicated, comments help to match e.g. the algorithmic steps to the code.

Reply:
I agree that there can be such cases, but IMHO such cases are rare. If the code can be refactored (e.g., by improve naming or splitting in multiple functions), so that *What* comments are not necessary anymore, the refactoring should be preferred. Also, if there is a complicated algorithm, the algorithm should be moved into a separate function so that it can be tested separately (and which should also be tested then).

Some suggestions for refactoring the code that make the comments redundant:

```python
        # global variables
        [....]

        # local variables
        [...]
```

Instead of splitting the function in sections using comments, separate functions could be defined.

```python
        # get all allocation points and required sizes in a list (one element per state)
        # of lists of tuples (location, size)

        state_allocation_list = []
        [...]
```

A [dataclass](https://docs.python.org/3/library/dataclasses.html) with a `location` and `size` field could be introduced to represent the allocation points. Feel free to add a type annotation to `state_allocation_list`, if you think that makes the content of the list more clear. ~I don't understand what is meant by `one element per state`.~  After looking at the code, I understand it now.

```python
        # sort the list for each state by descending sizes
        for entry in state_allocation_list:
            entry.sort(key=lambda x: x[1], reverse=True)
```

If a dataclass is used, the key could be `x.size`. A more descriptive name for `entry` could be helpful.

```python
        # create the actual slot with the max size for this element
        states_slot_list = list(map(create_slot_with_size, max_size_per_slot))
```

The call to `create_slot_with_size` with `max_size_per_slot` as argument provides the same information as the comment.

```python
        # map the locations to the corresponding memory slot
        for entry in state_allocation_list:
            assert len(entry) <= len(states_slot_list)
            for ((location, _), slot) in zip(entry, states_slot_list):
                map_location_to_slot(location, slot)
```

The call to `map_location_to_slot` provides the same information as the comment.

Reply:
I like your suggestions and have integrated them (I think - I have also done some other refactorings). I still disagree with your general stance against comments. This code needs some higher-level explanation, IMO. I added a docstring, though I can also transform it into a comment at the start of the function.
I agree the comments are redundant in a strict sense (convey the same information), but they really help understand the code in my opinion.
As an example for highly-commented code, you can check out the front-end sources of GNAT. There are "what"-style comments everywhere, and I believe they contribute to the clarity of the code.